sphinx_gallery/directives.py

```diff
@@ -144,21 +144,41 @@ def has_backrefs(obj):
                 gallery_dirs = config.sphinx_gallery_conf["gallery_dirs"]
                 if not isinstance(gallery_dirs, list):
                     gallery_dirs = [gallery_dirs]
-                dirs = [
-                    (e, g)
-                    for e, g in zip(examples_dirs, gallery_dirs)
-                    if (obj.find(e) != -1)
-                ]
-                if len(dirs) != 1:
+
+                gal_matches = []
+                for e, g in zip(examples_dirs, gallery_dirs):
+                    e = Path(src_dir, e).resolve(strict=True)
+                    g = Path(src_dir, g).resolve(strict=True)
+                    try:
+                        gal_matches.append((path.relative_to(e), g))
+                    except ValueError:
+                        continue
+                # `path` inside one `examples_dirs`
+                if (n_match := len(gal_matches)) == 1:
+                    ex_parents, target_dir = (
+                        gal_matches[0][0].parents,
+                        gal_matches[0][1],
+                    )
+                # `path` inside several `examples_dirs`, take the longest match
+                elif n_match > 1:
+                    ex_parents, target_dir = max(
+                        [(match[0].parents, match[1]) for match in gal_matches],
+                        key=lambda x: len(x[0]),
+                    )
+                # `path` is not inside a `examples_dirs`
+                else:
                     raise ExtensionError(
```

Comment:
This is the case where the glob match is not inside an `examples_dirs` (i.e., in old code when `len(dirs) = 0`).